help_to_heat/portal/download_views.py

```diff
@@ -326,3 +326,13 @@ def create_feedback_csv(feedbacks, file_name):
     full_file_name = f"feedback-data-{file_name}"
     response = create_csv(feedback_column_headings, rows, full_file_name)
     return response
+
+
+@require_http_methods(["GET"])
+@decorators.requires_service_manager
+def download_all_referrals(_request):
+    referrals = models.Referral.objects.all()
+    downloaded_at = timezone.now()
+    file_name = downloaded_at.strftime("%d-%m-%Y %H_%M")
+    response = create_referral_csv(referrals, file_name)
```

Comment:
From a quick look, I think this will blow up when any of `contact_number`, `country`, `epc_rating`, or `council_tax_band` is missing from `data`. There's some referrals in the prod DB with the last two missing, so I think we'll need to make sure our code is robust to this.

I think we can do that by adding defaults to the `dict.get` calls, so it doesn't default to `None`. Would you mind looking at doing that?

Please also double-check my working in getting those columns, and maybe have a look to see if there's any such referrals in dev or uat that we can test with?

Reply:
looking into this, I wasn't sure what suggests country or council_tax_band are necessary fields? deleting them from my local data seems to work okay, & I was under the impression only fields with logic applied (in this case string concatenation) in `add_extra_row_data` were pertinent

added defaults for `contact_number` & `uprn` by this logic, seems to export the csvs from the admin command fine enough

Reply:
I might be wrong - I expect you've got more Python experience than me - but I thought that evaluating `None in ("tuple", "of", "strings")` would error? We (potentially) do that with `country`, `epc_rating`, and `council_tax_band` inside `calculate_eligibility`. It might be that we never hit the problematic branches in the data you're testing on, because the function returns early?

`uprn` I think is safe because we have the `if uprn else ""`, but I think in general adding defaults across the board is sensible.

Reply:
for python this shouldn't error, since tuples aren't constrained to a single type so it is possible a tuple of strings contains None. in general when we designed calculate_eligibility it was done resiliently since we were keeping open the idea of calling it on an incomplete form